File: domain/signals/signal_buckets.py
from __future__ import annotations

from copy import deepcopy
from typing import Any, Literal


ShortTermSignalBucket = Literal["active", "waiting", "excluded"]

ACTIVE_BUCKET: ShortTermSignalBucket = "active"
WAITING_BUCKET: ShortTermSignalBucket = "waiting"
EXCLUDED_BUCKET: ShortTermSignalBucket = "excluded"

_ACTIONABLE_TRADE_STATES = {
    "ENTER NOW",
    "CONFIRMED BREAKOUT — BUY",
    "CONFIRMED BREAKDOWN — SELL",
}
_WAITING_TRADE_STATES = {
    "WAIT FOR PULLBACK",
    "BREAKOUT WATCH",
    "BREAKDOWN WATCH",
}
_WAITING_RECOMMENDATIONS = {
    "STRONG SETUP",
    "WATCHLIST",
}


def _normalized(value: Any) -> str:
    return str(value or "").strip().upper()


def classify_short_term_signal(row: dict[str, Any]) -> ShortTermSignalBucket:
    """Classify a short-term row by its current execution eligibility."""

    trade_state = _normalized(row.get("trade_state"))
    recommendation = _normalized(row.get("recommendation_label"))
    is_blocked = bool(str(row.get("action_block_reason") or "").strip())

    if (
        not is_blocked
        and row.get("is_actionable_now") is True
        and trade_state in _ACTIONABLE_TRADE_STATES
    ):
        return ACTIVE_BUCKET

    if is_blocked:
        return EXCLUDED_BUCKET

    if trade_state in _WAITING_TRADE_STATES or recommendation in _WAITING_RECOMMENDATIONS:
        return WAITING_BUCKET

    return EXCLUDED_BUCKET


def apply_short_term_signal_buckets(payload: dict[str, Any]) -> dict[str, Any]:
    """Attach a bucket to every short-term row and publish exact bucket counts."""

    protected = deepcopy(payload)
    counts = {
        ACTIVE_BUCKET: 0,
        WAITING_BUCKET: 0,
        EXCLUDED_BUCKET: 0,
    }
    rows: list[dict[str, Any]] = []

    for source_row in protected.get("short_term", []):
        if not isinstance(source_row, dict):
            continue
        row = dict(source_row)
        bucket = classify_short_term_signal(row)
        row["signal_bucket"] = bucket
        counts[bucket] += 1
        rows.append(row)

    stats = dict(protected.get("market_stats") or {})
    stats["actionable_now"] = counts[ACTIVE_BUCKET]
    stats["short_term_buckets"] = counts
    protected["market_stats"] = stats
    protected["short_term"] = rows
    return protected


__all__ = [
    "ACTIVE_BUCKET",
    "EXCLUDED_BUCKET",
    "ShortTermSignalBucket",
    "WAITING_BUCKET",
    "apply_short_term_signal_buckets",
    "classify_short_term_signal",
]

File: frontend/lib/signalBuckets.ts
import type { ApiRecord, ScanPayload } from "@/lib/api";


export type ShortTermSignalBucket = "active" | "waiting" | "excluded";

export const ACTIVE_BUCKET: ShortTermSignalBucket = "active";
export const WAITING_BUCKET: ShortTermSignalBucket = "waiting";
export const EXCLUDED_BUCKET: ShortTermSignalBucket = "excluded";

const ACTIONABLE_TRADE_STATES = new Set([
  "ENTER NOW",
  "CONFIRMED BREAKOUT — BUY",
  "CONFIRMED BREAKDOWN — SELL"
]);
const WAITING_TRADE_STATES = new Set([
  "WAIT FOR PULLBACK",
  "BREAKOUT WATCH",
  "BREAKDOWN WATCH"
]);
const WAITING_RECOMMENDATIONS = new Set([
  "STRONG SETUP",
  "WATCHLIST"
]);

function normalized(value: unknown): string {
  return String(value ?? "").trim().toUpperCase();
}

export function classifyShortTermSignal(row: ApiRecord): ShortTermSignalBucket {
  const tradeState = normalized(row.trade_state);
  const recommendation = normalized(row.recommendation_label);
  const isBlocked = String(row.action_block_reason ?? "").trim().length > 0;

  if (
    !isBlocked
    && row.is_actionable_now === true
    && ACTIONABLE_TRADE_STATES.has(tradeState)
  ) {
    return ACTIVE_BUCKET;
  }

  if (isBlocked) return EXCLUDED_BUCKET;

  if (
    WAITING_TRADE_STATES.has(tradeState)
    || WAITING_RECOMMENDATIONS.has(recommendation)
  ) {
    return WAITING_BUCKET;
  }

  return EXCLUDED_BUCKET;
}

export function applyShortTermSignalBuckets(payload: ScanPayload): ScanPayload {
  const counts: Record<ShortTermSignalBucket, number> = {
    active: 0,
    waiting: 0,
    excluded: 0
  };
  const rows = (Array.isArray(payload.short_term) ? payload.short_term : []).map((sourceRow) => {
    const row: ApiRecord = { ...sourceRow };
    const bucket = classifyShortTermSignal(row);
    row.signal_bucket = bucket;
    counts[bucket] += 1;
    return row;
  });

  return {
    ...payload,
    short_term: rows,
    market_stats: {
      ...(payload.market_stats ?? {}),
      actionable_now: counts.active,
      short_term_buckets: counts
    }
  };
}
